fidelius.py

```diff
@@ -0,0 +1,193 @@
+            encryption_request.sender_private_key,
+            encryption_request.requester_public_key
+        )
+        aes_encryption_key = cls.sha256_hkdf(salt, shared_secret, 32)
+        string_bytes = encryption_request.string_to_encrypt.encode('utf-8')
+
+        cipher = AES.new(aes_encryption_key, AES.MODE_GCM, iv)
+        encrypted_data, tag = cipher.encrypt_and_digest(string_bytes)
+        return base64.b64encode(encrypted_data + tag).decode('utf-8')
+
+    @classmethod
+    def decrypt(cls, decryption_request: DecryptionRequest):
+        sender_nonce = base64.b64decode(decryption_request.sender_nonce)
+        requester_nonce = base64.b64decode(decryption_request.requester_nonce)
+
+        # Calculate IV and salt from nonces
+        xor_of_nonces = bytes(
+            a ^ b for a, b in zip(sender_nonce, requester_nonce)
+        )
+        iv = xor_of_nonces[-12:]
+        salt = xor_of_nonces[:20]
+
+        shared_secret = cls.compute_shared_secret(
+            decryption_request.requester_private_key,
+            decryption_request.sender_public_key
+        )
+        aes_encryption_key = cls.sha256_hkdf(salt, shared_secret, 32)
+        encrypted_string = base64.b64decode(decryption_request.encrypted_data)[:-16]
+
+        cipher = AES.new(aes_encryption_key, AES.MODE_GCM, iv)
+        decrypted_string = cipher.decrypt(encrypted_string)
+
+        return decrypted_string.decode('utf-8')
+
+    @classmethod
+    def decode_base64_to_private_key(cls, encoded_key) -> int:
+        key_bytes = base64.b64decode(encoded_key.encode('utf-8'))
+        return int.from_bytes(key_bytes, byteorder='big')
+
+    @classmethod
+    def decode_base64_to_public_key(cls, encoded_key) -> Point:
+        key_bytes = base64.b64decode(encoded_key.encode('utf-8'))
+        # Ensure the key is in uncompressed form (starts with 0x04)
+        if key_bytes[0] != 0x04:
+            raise ValueError("Invalid public key format")
+        x_bytes = key_bytes[1:33]  # 32 bytes for x-coordinate
+        y_bytes = key_bytes[33:]    # 32 bytes for y-coordinate
+        x = int.from_bytes(x_bytes, byteorder='big')
+        y = int.from_bytes(y_bytes, byteorder='big')
+        return Point(x, y, curve=BC25519)
+
+    @classmethod
+    def compute_shared_secret(cls, sender_private_key, requester_public_key):
+        private_key_int = cls.decode_base64_to_private_key(sender_private_key)
+        public_key_point = cls.decode_base64_to_public_key(requester_public_key)
+        return KeyMaterial.encode_private_key_to_base64(
+            (private_key_int * public_key_point).x
+        )
+
+    @classmethod
+    def sha256_hkdf(cls, salt, shared_secret, key_length_in_bytes):
+        # Decode the initial key material from base64
+        shared_secret_bytes = base64.b64decode(shared_secret)
+        # Perform HKDF using SHA-256
+        hkdf = HKDF(shared_secret_bytes, key_length_in_bytes, salt, SHA256)
+        encryption_key = hkdf[:key_length_in_bytes]
```

Comment:
Curious if this required as hkdf should have  the desired length ?

Reply:
Good catch, that was remnant from previous attempt that I did, removed it now.